src/PersistentOrderedMap.mo

```diff
@@ -0,0 +1,892 @@
+  /// values in no specific order, or `null` when out of values to iterate over.
+  ///
+  /// Example:
+  /// ```motoko
+  /// import Map "mo:base/PersistentOrderedMap";
+  /// import Nat "mo:base/Nat"
+  /// import Iter "mo:base/Iter"
+  ///
+  /// let mapOps = Map.MapOps<Nat>(Nat.compare);
+  /// let rbMap = mapOps.fromIter<Text>(Iter.fromArray([(0, "Zero"), (2, "Two"), (1, "One")]));
+  ///
+  /// Debug.print(debug_show(Iter.toArray(Map.vals(rbMap))));
+  ///
+  /// // ["Zero", "One", "Two"]
+  /// ```
+  /// Cost of iteration over all elements:
+  /// Runtime: `O(n)`.
+  /// Space: `O(log(n))` retained memory plus garbage, see the note below.
+  /// where `n` denotes the number of key-value entries stored in the map.
+  ///
+  /// Note: Full map iteration creates `O(n)` temporary objects that will be collected as garbage.
+  public func vals<K, V>(m : Map<K, V>) : I.Iter<V>
+    = I.map(entries(m), func(kv : (K, V)) : V {kv.1});
+
+  /// Creates a new map by applying `f` to each entry in `rbMap`. Each entry
+  /// `(k, v)` in the old map is transformed into a new entry `(k, v2)`, where
+  /// the new value `v2` is created by applying `f` to `(k, v)`.
+  ///
+  /// Example:
+  /// ```motoko
+  /// import Map "mo:base/PersistentOrderedMap";
+  /// import Nat "mo:base/Nat"
+  /// import Iter "mo:base/Iter"
+  ///
+  /// let mapOps = Map.MapOps<Nat>(Nat.compare);
+  /// let rbMap = mapOps.fromIter<Text>(Iter.fromArray([(0, "Zero"), (2, "Two"), (1, "One")]));
+  ///
+  /// func f(key : Nat, _val : Text) : Nat = key * 2;
+  ///
+  /// let resMap = Map.map(rbMap, f);
+  ///
+  /// Debug.print(debug_show(Iter.toArray(Map.entries(resMap))));
+  ///
+  /// // [(0, 0), (1, 2), (2, 4)]
+  /// ```
+  ///
+  /// Cost of mapping all the elements:
+  /// Runtime: `O(n)`.
+  /// Space: `O(n)` retained memory
+  /// where `n` denotes the number of key-value entries stored in the map.
+  public func map<K, V1, V2>(rbMap : Map<K, V1>, f : (K, V1) -> V2) : Map<K, V2> {
+    func mapRec(m : Map<K, V1>) : Map<K, V2> {
+      switch m {
+        case (#leaf) { #leaf };
+        case (#node(c, l, xy, r)) {
+          #node(c, mapRec l, (xy.0, f xy), mapRec r) // TODO: try destination-passing style to avoid non tail-call recursion
+        };
+      }
+    };
+    mapRec(rbMap)
+  };
+
+  /// Determine the size of the tree as the number of key-value entries.
+  ///
+  /// Example:
+  /// ```motoko
+  /// import Map "mo:base/PersistentOrderedMap";
+  /// import Nat "mo:base/Nat"
+  /// import Iter "mo:base/Iter"
+  ///
+  /// let mapOps = Map.MapOps<Nat>(Nat.compare);
+  /// let rbMap = mapOps.fromIter<Text>(Iter.fromArray([(0, "Zero"), (2, "Two"), (1, "One")]));
+  ///
+  /// Debug.print(debug_show(Map.size(rbMap)));
+  ///
+  /// // 3
+  /// ```
+  ///
+  /// Runtime: `O(n)`.
+  /// Space: `O(1)`.
+  /// where `n` denotes the number of key-value entries stored in the tree.
+  public func size<K, V>(t : Map<K, V>) : Nat {
```

Comment:
I am not sure that it is sufficient to keep size in the root only for sets.
We have already started to work on them and added operations join and split for which it is not trivial to keep size only in the root.

We should think about it

Reply:
I think it's very likely possible to support it. It may introduce some overhead for keeping and passing through recursion not only a tree but also its size. 

For the current implementation, it may optimize `equal` and `isSubset`. While other operations like `intersect` don't have to know which tree operand is smaller.

So I am not sure, Is it worth it or not.. Maybe the user can do it on her own..

Reply:
Did you see the book by Nipkow? It has some suggestions on how to implement union etc  given maps. Not sure they apply but worth looking.

Reply:
Functional Data Structures and Algorithms
Tobias Nipkow, Jasmin Blanchette, Manuel Eberl, Alejandro Gómez-Londoño, Peter Lammich, Christian Sternagel, Simon Wimmer, Bohua Zhan
https://functional-algorithms-verified.org/

Reply:
>Did you see the book by Nipkow?

We actually took it from there. It uses `join`, `join2` etc which may have some constant overhead on supporting the size of intermediate trees without benefits to `union`, `intersect`, and `difference`.

`equal` and `isSubset` could have improvements for specific contention, but I am not sure that it is worth it.

Now you actually can check the draft of the 3rd Milestone with the implementation: 
https://github.com/dfinity/motoko-base/pull/662/files#diff-726841fb091c7380251085561293f301b9ee88b7a005ebd40038ccbe79cf96c4R227-R230

Reply:
I'll resolve this since you plan to implment 0(1) size in milestone 3.